bdk-ffi/src/esplora.rs

```diff
@@ -17,41 +23,57 @@ impl EsploraClient {
         Self(client)
     }
 
-    // This is a temporary solution for scanning. The long-term solution involves not passing
-    // the wallet to the client at all.
     pub fn full_scan(
         &self,
-        wallet: Arc<Wallet>,
+        request: Arc<FullScanRequest>,
         stop_gap: u64,
         parallel_requests: u64,
     ) -> Result<Arc<Update>, EsploraError> {
-        let wallet = wallet.get_wallet();
+        // using option and take is not ideal but the only way to take full ownership of the request
+        let request: BdkFullScanRequest<KeychainKind> = request
+            .0
+            .lock()
+            .unwrap()
+            .take()
+            .ok_or(EsploraError::RequestAlreadyConsumed)?;
 
-        let previous_tip = wallet.latest_checkpoint();
-        let keychain_spks = wallet.all_unbounded_spk_iters().into_iter().collect();
+        let result: BdkFullScanResult<KeychainKind> =
+            self.0
+                .full_scan(request, stop_gap as usize, parallel_requests as usize)?;
 
-        let (update_graph, last_active_indices) = self
-            .0
-            .full_scan(keychain_spks, stop_gap as usize, parallel_requests as usize)
-            .map_err(|e| EsploraError::from(*e))?;
+        let update = bdk::wallet::Update {
+            last_active_indices: result.last_active_indices,
+            graph: result.graph_update,
+            chain: Some(result.chain_update),
+        };
 
-        let missing_heights = update_graph.missing_heights(wallet.local_chain());
-        let chain_update = self
+        Ok(Arc::new(Update(update)))
+    }
+
+    pub fn sync(
+        &self,
+        request: Arc<SyncRequest>,
+        parallel_requests: u64,
+    ) -> Result<Arc<Update>, EsploraError> {
+        // using option and take is not ideal but the only way to take full ownership of the request
+        let request: BdkSyncRequest = request
             .0
-            .update_local_chain(previous_tip, missing_heights)
-            .map_err(|e| EsploraError::from(*e))?;
+            .lock()
+            .unwrap()
+            .take()
+            .ok_or(EsploraError::RequestAlreadyConsumed)?;
```

Comment:
Can be handled in a follow up PR, but what did ya'll think was the best way to handle this instead of using `unwrap()` which would panic?

Some idiomatic approaches I know of are:
- handle the option explicitly
- use match
- use if let

Handling the option explicitly could be something like:
```
let request = request
    .0
    .lock()
    .ok_or(EsploraError::LockError)?
    .take()
    .ok_or(EsploraError::RequestAlreadyConsumed)?;
```
... but one problem with that is we don't have an `EsploraError::LockError` and I'm not sure that error even fits as an EsploraError necessarily so if it was a different type we'd have to map it to an EsploraError since thats our return type, etc.

Thoughts on what is most idiomatic and what we use in our codebase that feels best?

Reply:
We have mutexes in a few other places, and at the moment we are using unwraps on all the lock methods. See [psbt.extract_tx()](https://github.com/bitcoindevkit/bdk-ffi/blob/7d951578d0b1b40a3f6f5a23b98166f72ae3845b/bdk-ffi/src/bitcoin.rs#L168-L172) and [DescriptorSecretKey.derive()](https://github.com/bitcoindevkit/bdk-ffi/blob/7d951578d0b1b40a3f6f5a23b98166f72ae3845b/bdk-ffi/src/keys.rs#L92).

It might be good to define a proper error for those, but I do imagine that would be maybe not within the scope of this update.

Reply:
> We have mutexes in a few other places, and at the moment we are using unwraps on all the lock methods. See [psbt.extract_tx()](https://github.com/bitcoindevkit/bdk-ffi/blob/7d951578d0b1b40a3f6f5a23b98166f72ae3845b/bdk-ffi/src/bitcoin.rs#L168-L172) and [DescriptorSecretKey.derive()](https://github.com/bitcoindevkit/bdk-ffi/blob/7d951578d0b1b40a3f6f5a23b98166f72ae3845b/bdk-ffi/src/keys.rs#L92).
> 
> It might be good to define a proper error for those, but I do imagine that would be maybe not within the scope of this update. Just my 2 cents.

Yeah, I mentioned it can be handled in a follow up PR because I wasn't wanting it to hold up anything about this PR (I'm ready to use all the goodies in this PR, heh), so I'm with ya there for sure. And thanks for calling out those other places for me like `extract_tx()` and `derive`